apiserver/common/leadership.go

```diff
@@ -103,47 +99,93 @@ func (a *leadershipPinningAPI) PinnedLeadership() (params.PinnedLeadershipResult
-	result := params.PinApplicationsResults{}
+// UnpinApplicationLeaders unpins leadership for applications based on the auth
+// tag provided.
+func (a *LeadershipPinning) UnpinApplicationLeaders() (params.PinApplicationsResults, error) {
+	if !a.authorizer.AuthMachineAgent() {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
 
 	tag := a.authorizer.GetAuthTag()
-	m, err := a.st.Machine(tag.Id())
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		return a.unpinMachineApplications(tag)
+	default:
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+}
+
+// GetMachineApplicationNames returns the applications associated with a
+// machine.
+func (a *LeadershipPinning) GetMachineApplicationNames(id string) ([]string, error) {
+	m, err := a.st.Machine(id)
 	if err != nil {
-		return result, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 	apps, err := m.ApplicationNames()
 	if err != nil {
-		return result, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
+	return apps, nil
+}
+
+// PinMachineApplicationsByName takes a slice of application names and attempts
+// to pin them accordingly.
+func (a *LeadershipPinning) PinMachineApplicationsByName(tag names.Tag, appNames []string) (params.PinApplicationsResults, error) {
+	return a.pinMachineAppsOps(tag, appNames, a.pinner.PinLeadership)
+}
+
+// UnpinMachineApplicationsByName takes a slice of application names and
+// attempts to pin them accordingly.
+func (a *LeadershipPinning) UnpinMachineApplicationsByName(tag names.Tag, appNames []string) (params.PinApplicationsResults, error) {
+	return a.pinMachineAppsOps(tag, appNames, a.pinner.UnpinLeadership)
+}
+
+// pinMachineApplications pins leadership for applications represented by units
+// running on the auth'd machine.
+func (a *LeadershipPinning) pinMachineApplications(tag names.Tag) (params.PinApplicationsResults, error) {
+	appNames, err := a.GetMachineApplicationNames(tag.Id())
+	if err != nil {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+	return a.pinMachineAppsOps(tag, appNames, a.pinner.PinLeadership)
+}
+
+// unpinMachineApplications unpins leadership for applications represented by
+// units running on the auth'd machine.
+func (a *LeadershipPinning) unpinMachineApplications(tag names.Tag) (params.PinApplicationsResults, error) {
+	appNames, err := a.GetMachineApplicationNames(tag.Id())
+	if err != nil {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+	return a.pinMachineAppsOps(tag, appNames, a.pinner.UnpinLeadership)
+}
+
+// pinMachineAppsOps runs the input pin/unpin operation against all
+// applications represented by units on the authorised machine.
+// An assumption is made that the validity of the auth tag has been verified
+// by the caller.
+func (a *LeadershipPinning) pinMachineAppsOps(tag names.Tag, appNames []string, op func(string, string) error) (params.PinApplicationsResults, error) {
```

Comment:
This now becomes `pinAppLeadersOps` or similar. Comment should be changed to reflect the less specific usage.